src/AggregateFunctions/AggregateFunctionGroupFormat.cpp

```diff
@@ -0,0 +1,238 @@
+    }
+
+    void add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        auto & data = this->data(place);
+        for (size_t i = 0; i < data.columns.size(); ++i)
+            data.columns[i]->insertFrom(*columns[i], row_num);
+    }
+
+    void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena *) const override
+    {
+        auto & data = this->data(place);
+        const auto & rhs_data = this->data(rhs);
+        for (size_t i = 0; i < data.columns.size(); ++i)
+            data.columns[i]->insertRangeFrom(*rhs_data.columns[i], 0, rhs_data.columns[i]->size());
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t>) const override
+    {
+        const auto & data = this->data(place);
+        UInt64 num_columns = data.columns.size();
+        UInt64 num_rows = num_columns ? data.columns.front()->size() : 0;
+
+        writeVarUInt(num_columns, buf);
+        writeVarUInt(num_rows, buf);
+
+        for (size_t i = 0; i < num_columns; ++i)
+        {
+            const auto & column = data.columns[i];
+            const auto & type = argument_types[i];
+            auto serialization = type->getDefaultSerialization();
+            NativeWriter::writeData(*serialization, column->getPtr(), buf, std::nullopt, 0, num_rows, DBMS_TCP_PROTOCOL_VERSION);
+        }
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, std::optional<size_t>, Arena *) const override
+    {
+        auto & data = this->data(place);
+        UInt64 num_columns = 0;
+        UInt64 num_rows = 0;
+        readVarUInt(num_columns, buf);
+        readVarUInt(num_rows, buf);
+
+        if (num_columns != argument_types.size())
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "State of aggregate function {} contains {} columns but {} columns are expected",
+                getName(),
+                num_columns,
+                argument_types.size());
+
+        data.columns.clear();
+        data.columns.reserve(num_columns);
+
+        for (size_t i = 0; i < num_columns; ++i)
+        {
+            const auto & type = argument_types[i];
+            auto serialization = type->getDefaultSerialization();
+            ColumnPtr column = type->createColumn();
+            NativeReader::readData(*serialization, column, buf, nullptr, num_rows, nullptr, nullptr);
+            data.columns.emplace_back(column->assumeMutable());
+        }
+    }
+
+    void insertResultInto(AggregateDataPtr __restrict place, IColumn & to, Arena *) const override
+    {
+        auto & data = this->data(place);
+        auto & column_string = assert_cast<ColumnString &>(to);
+
+        WriteBufferFromOwnString buffer;
+        auto output = FormatFactory::instance().getOutputFormat(format_name, buffer, header, context, format_settings);
```

Comment:
This should be initialized in the constructor and kept as a class member. We can "reset" the format for each invocation.

Reply:
The problem is that aggregate function instances are shared across threads, and WriteBuffer/IOutputFormat are stateful and not thread‑safe, so this would need locking and careful reset.